create-node-api/src/main/scala/com/normation/plugins/createnodeapi/CreateNodeApi.scala

```diff
@@ -0,0 +1,261 @@
+    def process0(version: ApiVersion, path: ApiPath, req: Req, params: DefaultParams, authzToken: AuthzToken): LiftResponse = {
+      (for {
+        json              <- req.json ?~! "This API only Accept JSON request"
+        nodes             <- Serialize.parseAll(json)
+      } yield {
+        import com.softwaremill.quicklens._
+        nodes.foldLeft(ResultHolder(Nil, Nil)) { case (res, node) =>
+          // now, try to save each node
+          saveNode(node) match {
+            case Right(id) => res.modify(_.created).using(_ :+id)
+            case Left(err) => res.modify(_.failed ).using(_ :+ ((node.id, err)) )
+          }
+        }
+      }) match {
+        case Full(resultHolder) =>
+          // if all succes, return success.
+          // Or if at least one is not failed, return success ?
+          if(resultHolder.failed.isEmpty) {
+            RestUtils.toJsonResponse(None, resultHolder.toJson)(schema.name, params.prettify)
+          } else {
+            RestUtils.toJsonError(None, resultHolder.toJson)(schema.name, params.prettify)
+          }
+        case eb: EmptyBox =>
+          val err = eb ?~! "Error when trying to parse node creation request"
+          RestUtils.toJsonError(None, JString(err.messageChain))(schema.name, params.prettify)
+      }
+    }
+
+    /// utility functions ///
+
+    /*
+     * for a given nodedetails, we:
+     * - try to convert it to inventory/node setup info
+     * - save inventory
+     * - if needed, accept
+     * - now, setup node info (property, state, etc)
+     */
+    def saveNode(nodeDetails: Rest.NodeDetails): Either[CreationError, NodeId] = {
+      import Validated._
+      def toCreationError(res: ValidatedNel[NodeValidationError, NodeTemplate]) = {
+        res match {
+          case Invalid(nel) => Left(CreationError.OnValidation(nel))
+          case Valid(r)     => Right(r)
+        }
+      }
+
+      for {
+        validated <- toCreationError(Validation.toNodeTemplate(nodeDetails))
+        created   <- saveInventory(validated.inventory)
+        nodeSetup <- accept(validated)
+        nodeId    <- saveRudderNode(validated.inventory.node.main.id, nodeSetup)
+      } yield {
+        nodeId
+      }
+    }
+
+    /*
+     * Save the inventory part. Alway save in "pending", acceptation
+     * is done aftrward if needed.
+     */
+    def saveInventory(inventory: FullInventory): Either[CreationError, NodeId] = {
+      (for {
+        iSaved <- inventoryRepos.save(inventory)
+      } yield {
+        inventory.node.main.id
+      }) match {
+        case Full(id)     => Right(id)
+        case eb: EmptyBox => Left(CreationError.OnSaveInventory((eb?~!"Error during node creation").messageChain))
+      }
+    }
+
+    def accept(template: NodeTemplate): Either[CreationError, NodeSetup] = {
+      val id = template.inventory.node.main.id
+      newNodeManager.accept(id, ModificationId(uuidGen.newUuid), Rest.CreateNodeActor) match {
```

Comment:
Don't we want to pass the actor calling the api ? identified through it's api token ?

Reply:
yes, certainly so